microceph/ceph/nfs.go

```diff
@@ -0,0 +1,287 @@
+		if err != nil {
+			logger.Errorf("Cleaning up '%s' failed: %v", ganeshaConfDir, err)
+		}
+	})
+
+	// Create NFS Ganesha configuration.
+	userID := fmt.Sprintf("nfs.%s.%s", nfs.ClusterID, hostname)
+	configs := map[string]any{
+		"bindAddr":      nfs.BindAddress,
+		"bindPort":      nfs.BindPort,
+		"snapDir":       pathConsts.SnapPath,
+		"runDir":        pathConsts.RunPath,
+		"confDir":       ganeshaConfDir,
+		"userID":        userID,
+		"clusterID":     nfs.ClusterID,
+		"minorVersions": nfs.V4MinVersion,
+	}
+
+	ganeshaConf := newGaneshaConfig(ganeshaConfDir)
+	err = ganeshaConf.WriteConfig(configs, 0644)
+	if err != nil {
+		return err
+	}
+
+	// Create NFS Ganesha Ceph configuration.
+	configs = map[string]any{
+		"confDir":  ganeshaConfDir,
+		"monitors": strings.Join(monitorAddresses, ","),
+	}
+
+	cephConf := newGaneshaCephConfig(ganeshaConfDir)
+	err = cephConf.WriteConfig(configs, 0644)
+	if err != nil {
+		return err
+	}
+
+	// Create NFS Ganesha Ceph keyring.
+	logger.Debugf("Creating ceph client 'client.%s' for NFS Ganesha (ClusterID '%s')", userID, nfs.ClusterID)
+	err = createNFSKeyring(ganeshaConfDir, nfs.ClusterID, userID)
+	if err != nil {
+		return err
+	}
+
+	revert.Add(func() {
+		err := DeleteClientKey(userID)
+		if err != nil {
+			logger.Errorf("Cleaning up NFS Ganesha ceph client 'client.%s' failed: %v", userID, err)
+		}
+	})
+
+	// Create the NFS Pools if needed.
+	logger.Debugf("Creating NFS Rados Pools (ClusterID '%s')", nfs.ClusterID)
+	err = ensureNFSPools(nfs.ClusterID)
+	if err != nil {
+		return err
+	}
+
+	// Add the node to the Shared Grace Management Database.
+	logger.Debugf("Adding node to Shared Grace Management Database (ClusterID '%s')", nfs.ClusterID)
+	err = addNodeToSharedGraceMgmtDb(filepath.Join(ganeshaConfDir, "ceph.conf"), nfs.ClusterID, userID, hostname)
+	if err != nil {
+		return err
+	}
+
+	revert.Add(func() {
+		err := removeNodeFromSharedGraceMgmtDb(filepath.Join(ganeshaConfDir, "ceph.conf"), nfs.ClusterID, userID, hostname)
+		if err != nil {
+			logger.Errorf("Removing node from Shared Grace Management Database failed: %v", err)
+		}
+	})
+
+	// Start the NFS Ganesha service.
+	logger.Debugf("Starting NFS Ganesha service (ClusterID '%s')", nfs.ClusterID)
+	err = startNFS()
+	if err != nil {
+		return err
+	}
+
+	revert.Success() // Added revert functions are not run on return.
+
+	logger.Debugf("Enabled NFS on node with ClusterID '%s'", nfs.ClusterID)
+
+	return nil
+}
+
+// DisableNFS disables the NFS service on the cluster.
+func DisableNFS(ctx context.Context, s interfaces.StateInterface, clusterID string) error {
+	logger.Debugf("Disabling NFS on node with ClusterID '%s'", clusterID)
+	hostname, err := os.Hostname()
+	if err != nil {
+		return err
+	}
+
+	pathConsts := constants.GetPathConst()
+	ganeshaConfDir := filepath.Join(pathConsts.ConfPath, "ganesha")
+
+	// Stop the NFS Ganesha service.
+	logger.Debugf("Stopping NFS Ganesha service (ClusterID '%s')", clusterID)
```

Comment:
The method `DisableNFS` does not verify whether the nfs daemon runnning on the host belongs to the provided cluster-id or not. This may cause incorrect service disablement if an operator (by mistake) passes the wrong cluster-id on the wrong host.